File: bot/bot.py
import os
import json
import time
import threading
import html
from pathlib import Path

import telebot
from telebot import types
from dotenv import load_dotenv


# ============================================================
# ENV
# ============================================================

load_dotenv()

BOT_TOKEN = os.getenv("BOT_TOKEN")

if not BOT_TOKEN:
    raise RuntimeError("BOT_TOKEN is missing from .env")


# ============================================================
# PATHS
# ============================================================

BASE_DIR = Path(__file__).resolve().parent

SUBLINK_PATH = (
    BASE_DIR
    / ".."
    / "crawler"
    / "sublink.txt"
).resolve()

USERS_PATH = BASE_DIR / "users.json"


# ============================================================
# BOT
# ============================================================

bot = telebot.TeleBot(
    BOT_TOKEN,
    parse_mode=None
)


# ============================================================
# USERS
# ============================================================

def load_users():

    try:

        if not USERS_PATH.exists():

            USERS_PATH.write_text(
                "[]",
                encoding="utf-8"
            )

            return []

        with open(
            USERS_PATH,
            "r",
            encoding="utf-8"
        ) as file:

            data = json.load(file)

        if not isinstance(data, list):
            return []

        return data

    except Exception as error:

        print(
            "Failed to load users:",
            error
        )

        return []


def save_users(users):

    with open(
        USERS_PATH,
        "w",
        encoding="utf-8"
    ) as file:

        json.dump(
            users,
            file,
            indent=4
        )


def add_user(chat_id):

    users = load_users()

    if chat_id not in users:

        users.append(chat_id)

        save_users(users)

        print(
            f"New user added: {chat_id}"
        )


# ============================================================
# LOAD CONFIGS
# ============================================================

def load_configs():

    try:

        with open(
            SUBLINK_PATH,
            "r",
            encoding="utf-8"
        ) as file:

            text = file.read()

        return [
            line.strip()
            for line in text.splitlines()
            if line.strip()
        ]

    except Exception as error:

        print(
            "Failed to load sublink.txt:",
            error
        )

        return []


def load_sublink():

    try:

        with open(
            SUBLINK_PATH,
            "r",
            encoding="utf-8"
        ) as file:

            return file.read().strip()

    except Exception as error:

        print(
            "Failed to load sublink.txt:",
            error
        )

        return ""


# ============================================================
# MAIN KEYBOARD
# ============================================================

def main_keyboard():

    keyboard = types.InlineKeyboardMarkup()

    subscription_button = types.InlineKeyboardButton(
        "🔗 Subscription Link",
        callback_data="subscription"
    )

    configs_button = types.InlineKeyboardButton(
        "⚙️ Configs",
        callback_data="configs"
    )

    keyboard.row(
        subscription_button
    )

    keyboard.row(
        configs_button
    )

    return keyboard


# ============================================================
# /START
# ============================================================

@bot.message_handler(
    commands=["start"]
)
def start_command(message):

    chat_id = message.chat.id

    print(
        f"START received from {chat_id}"
    )

    add_user(chat_id)

    try:

        bot.send_message(
            chat_id,
            "Welcome to ConfigHub.\n\n"
            "Choose what you need:",
            reply_markup=main_keyboard()
        )

        print(
            f"START response sent to {chat_id}"
        )

    except Exception as error:

        print(
            f"START ERROR for {chat_id}:",
            error
        )


# ============================================================
# SUBSCRIPTION LINK
# ============================================================

SUBSCRIPTION_URL = "https://qrco.de/bgy6Bk"


@bot.callback_query_handler(
    func=lambda call:
        call.data == "subscription"
)
def subscription_callback(call):

    try:

        bot.answer_callback_query(
            call.id
        )

        bot.send_message(
            call.message.chat.id,
            SUBSCRIPTION_URL
        )

    except Exception as error:

        print(
            "SUBSCRIPTION ERROR:",
            error
        )
# ============================================================
# CONFIG LIST
# ============================================================

@bot.callback_query_handler(
    func=lambda call:
        call.data == "configs"
)
def configs_callback(call):

    try:

        bot.answer_callback_query(
            call.id
        )

        configs = load_configs()

        if not configs:

            bot.send_message(
                call.message.chat.id,
                "❌ No configurations are currently available."
            )

            return

        keyboard = types.InlineKeyboardMarkup()

        # COPY ALL
        copy_all_button = types.InlineKeyboardButton(
            "📋 Copy All Configs",
            callback_data="copy_all"
        )

        keyboard.row(
            copy_all_button
        )

        # INDIVIDUAL CONFIGS
        for index, config in enumerate(configs):

            button = types.InlineKeyboardButton(
                f"⚡ Config {index + 1:03d}",
                callback_data=f"config:{index}"
            )

            keyboard.row(
                button
            )

        back_button = types.InlineKeyboardButton(
            "⬅️ Back",
            callback_data="main"
        )

        keyboard.row(
            back_button
        )

        bot.send_message(
            call.message.chat.id,
            "Choose a configuration:",
            reply_markup=keyboard
        )

    except Exception as error:

        print(
            "CONFIG LIST ERROR:",
            error
        )


# ============================================================
# COPY ALL CONFIGS
# ============================================================

@bot.callback_query_handler(
    func=lambda call:
        call.data == "copy_all"
)
def copy_all_callback(call):

    try:

        bot.answer_callback_query(
            call.id,
            "Sending all configurations..."
        )

        sublink = load_sublink()

        if not sublink:

            bot.send_message(
                call.message.chat.id,
                "❌ No configurations are currently available."
            )

            return

        # Telegram message limit is 4096 characters.
        # If the whole file fits, send it as a code block.
        if len(sublink) <= 4000:

            escaped = html.escape(sublink)

            message = (
                "<pre>"
                + escaped
                + "</pre>"
            )

            bot.send_message(
                call.message.chat.id,
                message,
                parse_mode="HTML"
            )

        else:

            # If it is too large, send the actual file.
            with open(
                SUBLINK_PATH,
                "rb"
            ) as file:

                bot.send_document(
                    call.message.chat.id,
                    file,
                    caption="📋 All configurations"
                )

    except Exception as error:

        print(
            "COPY ALL ERROR:",
            error
        )


# ============================================================
# SEND CONFIG
# ============================================================

@bot.callback_query_handler(
    func=lambda call:
        call.data.startswith("config:")
)
def config_callback(call):

    try:

        bot.answer_callback_query(
            call.id
        )

        index = int(
            call.data.split(":")[1]
        )

        configs = load_configs()

        if (
            index < 0
            or index >= len(configs)
        ):

            bot.send_message(
                call.message.chat.id,
                "❌ This configuration no longer exists."
            )

            return

        config = configs[index]

        # EXACT CONFIG
        bot.send_message(
            call.message.chat.id,
            config
        )

    except Exception as error:

        print(
            "CONFIG SEND ERROR:",
            error
        )


# ============================================================
# BACK TO MAIN
# ============================================================

@bot.callback_query_handler(
    func=lambda call:
        call.data == "main"
)
def main_callback(call):

    try:

        bot.answer_callback_query(
            call.id
        )

        bot.send_message(
            call.message.chat.id,
            "Choose what you need:",
            reply_markup=main_keyboard()
        )

    except Exception as error:

        print(
            "MAIN MENU ERROR:",
            error
        )


# ============================================================
# BROADCAST
# ============================================================

def broadcast_update():

    users = load_users()

    if not users:

        print(
            "No users to broadcast."
        )

        return

    print(
        f"Broadcasting update to {len(users)} users..."
    )

    message = (
        "🔄 Configurations updated!\n\n"
        "New configurations are available.\n\n"
        "Please update your Subscription."
    )

    invalid_users = []

    for chat_id in users:

        try:

            bot.send_message(
                chat_id,
                message,
                reply_markup=main_keyboard()
            )

            print(
                f"Broadcast sent to {chat_id}"
            )

        except Exception as error:

            print(
                f"Failed to send to {chat_id}:",
                error
            )

            error_text = str(error).lower()

            if (
                "bot was blocked" in error_text
                or "chat not found" in error_text
                or "user is deactivated" in error_text
            ):

                invalid_users.append(
                    chat_id
                )

        time.sleep(0.05)

    if invalid_users:

        remaining_users = [
            user
            for user in users
            if user not in invalid_users
        ]

        save_users(
            remaining_users
        )

    print(
        "Broadcast finished."
    )


# ============================================================
# SUBLINK WATCHER
# ============================================================

last_sublink_content = None


def initialize_sublink():

    global last_sublink_content

    try:

        with open(
            SUBLINK_PATH,
            "r",
            encoding="utf-8"
        ) as file:

            last_sublink_content = file.read()

        print(
            "Initial sublink loaded."
        )

        print(
            f"Watching: {SUBLINK_PATH}"
        )

    except Exception as error:

        print(
            "Could not read sublink.txt:",
            error
        )


def check_sublink():

    global last_sublink_content

    while True:

        try:

            with open(
                SUBLINK_PATH,
                "r",
                encoding="utf-8"
            ) as file:

                current_content = file.read()

            if last_sublink_content is None:

                last_sublink_content = (
                    current_content
                )

            elif (
                current_content
                != last_sublink_content
            ):

                print(
                    "================================"
                )

                print(
                    "sublink.txt changed!"
                )

                print(
                    "================================"
                )

                last_sublink_content = (
                    current_content
                )

                broadcast_update()

        except Exception as error:

            print(
                "Sublink check failed:",
                error
            )

        time.sleep(30)


# ============================================================
# INLINE MODE
# ============================================================

@bot.inline_handler(
    func=lambda query: True
)
def inline_query(query):

    try:

        configs = load_configs()

        results = []

        search = query.query.strip().lower()

        # ----------------------------------------------------
        # ALL CONFIGS
        # ----------------------------------------------------

        if (
            not search
            or search in ["all", "*"]
        ):

            all_configs = "\n".join(configs)

            if len(all_configs) <= 4096:

                results.append(
                    types.InlineQueryResultArticle(
                        id="all_configs",
                        title="📋 All Configs",
                        description=(
                            f"{len(configs)} configurations"
                        ),
                        input_message_content=
                        types.InputTextMessageContent(
                            all_configs
                        ),
                        reply_markup=None
                    )
                )

        # ----------------------------------------------------
        # INDIVIDUAL CONFIGS
        # ----------------------------------------------------

        for index, config in enumerate(configs):

            # Search inside config
            if search and search not in config.lower():
                continue

            preview = config

            if len(preview) > 80:
                preview = preview[:80] + "..."

            results.append(
                types.InlineQueryResultArticle(
                    id=f"config_{index}",
                    title=(
                        f"⚡ Config {index + 1:03d}"
                    ),
                    description=preview,
                    input_message_content=
                    types.InputTextMessageContent(
                        config
                    )
                )
            )

            # Telegram inline result limit
            if len(results) >= 50:
                break

        bot.answer_inline_query(
            query.id,
            results,
            cache_time=5,
            is_personal=True
        )

    except Exception as error:

        print(
            "INLINE ERROR:",
            error
        )

        try:

            bot.answer_inline_query(
                query.id,
                [],
                cache_time=1
            )

        except Exception:
            pass


# ============================================================
# START
# ============================================================

print(
    "================================"
)

print(
    "Starting ConfigHub bot..."
)

print(
    f"Watching: {SUBLINK_PATH}"
)

print(
    "================================"
)

initialize_sublink()


watcher_thread = threading.Thread(
    target=check_sublink,
    daemon=True
)

watcher_thread.start()


try:

    print(
        "Bot polling started."
    )

    bot.infinity_polling(
        skip_pending=True,
        timeout=30,
        long_polling_timeout=30
    )

except Exception as error:

    print(
        "BOT ERROR:",
        error
    )

File: bot/bot.ts
import { Bot, InlineKeyboard, InputFile } from "grammy";
import type {
  D1Database,
  ExecutionContext,
  ScheduledController,
} from "@cloudflare/workers-types";

interface Env {
  BOT_TOKEN: string;
  confighub_users: D1Database;
}

const SUBLINK_URL =
  "https://raw.githubusercontent.com/aghrabooti/V2ray_crawler/refs/heads/main/crawler/sublink.txt";

const SUBSCRIPTION_URL =
  "https://www.canvaqr.com/RGQotQsltm";


// ============================================================
// BOT
// ============================================================

async function createBot(env: Env) {
  const bot = new Bot(env.BOT_TOKEN);

  await bot.init();

  return bot;
}


// ============================================================
// USERS
// ============================================================

async function loadUsers(env: Env): Promise<string[]> {
  try {
    const result = await env.confighub_users
      .prepare("SELECT chat_id FROM users")
      .all<{ chat_id: string }>();

    return result.results.map((row) => row.chat_id);
  } catch (error) {
    console.error("Failed to load users:", error);
    return [];
  }
}


async function addUser(
  env: Env,
  chatId: string
): Promise<void> {
  try {
    await env.confighub_users
      .prepare(
        "INSERT OR IGNORE INTO users (chat_id) VALUES (?)"
      )
      .bind(chatId)
      .run();

    console.log(`User registered: ${chatId}`);
  } catch (error) {
    console.error("Failed to add user:", error);
  }
}


async function removeUser(
  env: Env,
  chatId: string
): Promise<void> {
  try {
    await env.confighub_users
      .prepare("DELETE FROM users WHERE chat_id = ?")
      .bind(chatId)
      .run();
  } catch (error) {
    console.error("Failed to remove user:", error);
  }
}


// ============================================================
// LOAD SUBLINK
// ============================================================

async function loadSublink(): Promise<string> {
  try {
    const response = await fetch(
      `${SUBLINK_URL}?t=${Date.now()}`,
      {
        headers: {
          "Cache-Control": "no-cache",
        },
      }
    );

    if (!response.ok) {
      throw new Error(
        `HTTP ${response.status}`
      );
    }

    return await response.text();
  } catch (error) {
    console.error(
      "Failed to load sublink:",
      error
    );

    return "";
  }
}


// ============================================================
// LOAD CONFIGS
// ============================================================

async function loadConfigs(): Promise<string[]> {
  const sublink = await loadSublink();

  if (!sublink) {
    return [];
  }

  return sublink
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
}


// ============================================================
// MAIN KEYBOARD
// ============================================================

function mainKeyboard() {
  return new InlineKeyboard()
    .text(
      "🔗 Subscription Link",
      "subscription"
    )
    .row()
    .text(
      "⚙️ Configs",
      "configs"
    );
}


// ============================================================
// /START
// ============================================================

function registerHandlers(
  bot: Bot,
  env: Env
) {

  bot.command(
    "start",
    async (ctx) => {

      try {

        const chatId = String(
          ctx.chat.id
        );

        console.log(
          `START received from ${chatId}`
        );

        await addUser(
          env,
          chatId
        );

        await ctx.reply(
          "Welcome to ConfigHub.\n\n" +
          "Choose what you need:",
          {
            reply_markup:
              mainKeyboard(),
          }
        );

      } catch (error) {

        console.error(
          "START ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // SUBSCRIPTION
  // ==========================================================

  bot.callbackQuery(
    "subscription",
    async (ctx) => {

      try {

        await ctx.answerCallbackQuery();

        await ctx.reply(
          SUBSCRIPTION_URL
        );

      } catch (error) {

        console.error(
          "SUBSCRIPTION ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // CONFIG LIST
  // ==========================================================

  bot.callbackQuery(
    "configs",
    async (ctx) => {

      try {

        await ctx.answerCallbackQuery();

        const configs =
          await loadConfigs();

        if (!configs.length) {

          await ctx.reply(
            "❌ No configurations are currently available."
          );

          return;
        }


        const keyboard =
          new InlineKeyboard();


        // COPY ALL

        keyboard
          .text(
            "📋 Copy All Configs",
            "copy_all"
          )
          .row();


        // INDIVIDUAL CONFIGS

        configs.forEach(
          (_, index) => {

            keyboard
              .text(
                `⚡ Config ${String(
                  index + 1
                ).padStart(3, "0")}`,
                `config:${index}`
              )
              .row();

          }
        );


        // BACK

        keyboard.text(
          "⬅️ Back",
          "main"
        );


        await ctx.reply(
          "Choose a configuration:",
          {
            reply_markup:
              keyboard,
          }
        );

      } catch (error) {

        console.error(
          "CONFIG LIST ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // COPY ALL
  // ==========================================================

  bot.callbackQuery(
    "copy_all",
    async (ctx) => {

      try {

        await ctx.answerCallbackQuery(
          "Sending all configurations..."
        );


        const sublink =
          await loadSublink();


        if (!sublink) {

          await ctx.reply(
            "❌ No configurations are currently available."
          );

          return;
        }


        // Telegram message limit

        if (sublink.length <= 4000) {

          const message =
            `<pre>${escapeHtml(
              sublink
            )}</pre>`;


          await ctx.reply(
            message,
            {
              parse_mode: "HTML",
            }
          );

        } else {

          const blob =
            new Blob(
              [sublink],
              {
                type: "text/plain",
              }
            );


          await ctx.replyWithDocument(
            new InputFile(
              blob,
              "sublink.txt"
            ),
            {
              caption:
                "📋 All configurations",
            }
          );

        }

      } catch (error) {

        console.error(
          "COPY ALL ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // INDIVIDUAL CONFIG
  // ==========================================================

  bot.callbackQuery(
    /^config:(\d+)$/,
    async (ctx) => {

      try {

        await ctx.answerCallbackQuery();


        const index =
          Number(
            ctx.match[1]
          );


        const configs =
          await loadConfigs();


        if (
          index < 0 ||
          index >= configs.length
        ) {

          await ctx.reply(
            "❌ This configuration no longer exists."
          );

          return;
        }


        await ctx.reply(
          configs[index]
        );

      } catch (error) {

        console.error(
          "CONFIG SEND ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // BACK
  // ==========================================================

  bot.callbackQuery(
    "main",
    async (ctx) => {

      try {

        await ctx.answerCallbackQuery();

        await ctx.reply(
          "Choose what you need:",
          {
            reply_markup:
              mainKeyboard(),
          }
        );

      } catch (error) {

        console.error(
          "MAIN MENU ERROR:",
          error
        );

      }

    }
  );


  // ==========================================================
  // INLINE MODE
  // ==========================================================

  bot.on(
    "inline_query",
    async (ctx) => {

      try {

        const configs =
          await loadConfigs();


        const results: any[] =
          [];


        const search =
          ctx.inlineQuery.query
            .trim()
            .toLowerCase();


        // ------------------------------------------------------
        // ALL CONFIGS
        // ------------------------------------------------------

        if (
          !search ||
          search === "all" ||
          search === "*"
        ) {

          const allConfigs =
            configs.join("\n");


          if (
            allConfigs.length <= 4096
          ) {

            results.push({

              type: "article",

              id: "all_configs",

              title:
                "📋 All Configs",

              description:
                `${configs.length} configurations`,

              input_message_content: {

                message_text:
                  allConfigs,

              },

            });

          }

        }


        // ------------------------------------------------------
        // INDIVIDUAL CONFIGS
        // ------------------------------------------------------

        for (
          let index = 0;
          index < configs.length;
          index++
        ) {

          const config =
            configs[index];


          if (
            search &&
            !config
              .toLowerCase()
              .includes(search)
          ) {

            continue;

          }


          let preview =
            config;


          if (
            preview.length > 80
          ) {

            preview =
              preview.slice(
                0,
                80
              ) + "...";

          }


          results.push({

            type: "article",

            id:
              `config_${index}`,

            title:
              `⚡ Config ${String(
                index + 1
              ).padStart(3, "0")}`,

            description:
              preview,

            input_message_content: {

              message_text:
                config,

            },

          });


          if (
            results.length >= 50
          ) {

            break;

          }

        }


        await ctx.answerInlineQuery(
          results,
          {
            cache_time: 5,
            is_personal: true,
          }
        );

      } catch (error) {

        console.error(
          "INLINE ERROR:",
          error
        );


        try {

          await ctx.answerInlineQuery(
            [],
            {
              cache_time: 1,
            }
          );

        } catch {}

      }

    }
  );

}


// ============================================================
// BROADCAST
// ============================================================

async function broadcastUpdate(
  env: Env
): Promise<void> {

  const bot =
    await createBot(env);


  const users =
    await loadUsers(env);


  if (!users.length) {

    console.log(
      "No users to broadcast."
    );

    return;

  }


  console.log(
    `Broadcasting update to ${users.length} users...`
  );


  const message =
    "🔄 Configurations updated!\n\n" +
    "New configurations are available.\n\n" +
    "Please update your Subscription.";


  for (
    const chatId of users
  ) {

    try {

      await bot.api.sendMessage(
        chatId,
        message,
        {
          reply_markup:
            mainKeyboard(),
        }
      );


      console.log(
        `Broadcast sent to ${chatId}`
      );

    } catch (error) {

      console.error(
        `Failed to send to ${chatId}:`,
        error
      );


      const errorText =
        String(error).toLowerCase();


      if (
        errorText.includes(
          "bot was blocked"
        ) ||
        errorText.includes(
          "chat not found"
        ) ||
        errorText.includes(
          "user is deactivated"
        )
      ) {

        await removeUser(
          env,
          chatId
        );

      }

    }

  }


  console.log(
    "Broadcast finished."
  );

}


// ============================================================
// HTML ESCAPE
// ============================================================

function escapeHtml(
  text: string
): string {

  return text
    .replace(
      /&/g,
      "&amp;"
    )
    .replace(
      /</g,
      "&lt;"
    )
    .replace(
      />/g,
      "&gt;"
    )
    .replace(
      /"/g,
      "&quot;"
    );

}


// ============================================================
// CLOUDFLARE WORKER
// ============================================================

export default {

  // ==========================================================
  // SCHEDULED
  // ==========================================================

  async scheduled(
    controller: ScheduledController,
    env: Env,
    ctx: ExecutionContext
  ): Promise<void> {

    console.log(
      `Scheduled event: ${new Date().toISOString()}`
    );

    // Nothing needs to run here.
    // The bot is webhook-based.
    //
    // GitHub sends POST /github-update
    // whenever crawler/sublink.txt changes.

  },


  // ==========================================================
  // FETCH
  // ==========================================================

  async fetch(
    request: Request,
    env: Env,
    ctx: ExecutionContext
  ): Promise<Response> {

    const url =
      new URL(request.url);


    // ========================================================
    // GITHUB WEBHOOK
    // ========================================================

    if (
      url.pathname ===
      "/github-update"
    ) {

      if (
        request.method !== "POST"
      ) {

        return new Response(
          "Method Not Allowed",
          {
            status: 405,
          }
        );

      }


      try {

        const body =
          await request.json() as {
            commits?: Array<{
              added?: string[];
              modified?: string[];
              removed?: string[];
            }>;
          };


        const commits =
          body.commits || [];


        const changed =
          commits.some(
            (commit) => {

              const files = [

                ...(commit.added || []),

                ...(commit.modified || []),

                ...(commit.removed || []),

              ];


              return files.some(
                (file) =>
                  file ===
                  "crawler/sublink.txt"
              );

            }
          );


        console.log(
          `GitHub webhook received. Changed: ${changed}`
        );


        if (changed) {

          // Wait for broadcast before
          // returning the response.

          await broadcastUpdate(
            env
          );

        }


        return Response.json({

          success: true,

          changed,

        });

      } catch (error) {

        console.error(
          "GITHUB WEBHOOK ERROR:",
          error
        );


        return Response.json(
          {
            success: false,
            error: "Invalid webhook payload",
          },
          {
            status: 400,
          }
        );

      }

    }


    // ========================================================
    // TELEGRAM WEBHOOK
    // ========================================================

    if (
      url.pathname ===
      "/telegram"
    ) {

      if (
        request.method !== "POST"
      ) {

        return new Response(
          "Method Not Allowed",
          {
            status: 405,
          }
        );

      }


      try {

        const bot =
          await createBot(env);


        registerHandlers(
          bot,
          env
        );


        const update =
          await request.json();


        await bot.handleUpdate(
          update as any
        );


        return new Response(
          "OK"
        );

      } catch (error) {

        console.error(
          "TELEGRAM WEBHOOK ERROR:",
          error
        );


        return new Response(
          "Error",
          {
            status: 500,
          }
        );

      }

    }


    // ========================================================
    // HEALTH CHECK
    // ========================================================

    return new Response(
      "ConfigHub bot is running."
    );

  },

};